README.md
# **modelica_bridge**
## Introduction
***modelica_bridge*** is the [ROS](www.ros.org) package complement of the [Modelica](www.modelica.org) package [***ROS_Bridge***](https://github.com/ModROS/ROS_Bridge.git).

The purpose of ***modelica_bridge*** is to provide an interface between Modelica tools and ROS. This interface, or bridge, is accomplished by running tcp/ip sockets in a ROS node, and in an external C-function through Modelica (accomplished via ***ROS_Bridge***).

## Usage Summary

The modelica_bridge package consists of the node connecting ROS and Modelica, **modbridge_node**, and the message type **modbridge_node** publishes on, the **ModComm.msg**. 

The package can take up to 256 inputs from the ROS controllers, and can send a maximum of 1024 characters over a character buffer to Modelica. It has two parameters: *port_num* and *update_rate*. Ensure that the *port_num* matches the value of the Modelica model's port number parameter.

For more detailed documentation and tutorials, see the [ROS wiki](wiki.ros.org/modelica_bridge).

## **modelica_bridge** Development Summary
### 0.1.1 (2018-07-02)
- Fixes issues with documentation
- Fixes issue of *CMakeLists.txt* not having install targets
- Cleans up and resolves *package.xml* dependencies list
#
### 0.1.0 (2018-06-30)
- Indexed and released to ros-buildfarm using **bloom**
- Added [wiki for **modelica_bridge**](wiki.ros.org/modelica_bridge)
- Added Doxygen documentation using rosdoc_lite
#
### 0.0.0 (2018-06-30)
- First full commit of the **modelica_bridge** package.
  - Previously known as **modros**

CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package modelica_bridge
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
0.1.1 (2018-07-02)
------------------
* Edited package description
* Updated documentation
* Edited README.md
* Edited documentation for modelica_bridge
* Edited <depend> tags in package.xml
* Added install target for modbridge node
* Edited documentation for modelica_bridge

0.1.0 (2018-06-30)
------------------
* Added link to online documentation of modelica_bridge
* Fixed CHANGELOG.rst
* Added Doxygen documentation generated by rosdoc_lite

0.0.0 (2018-06-30)
------------------
* First commit of full modelica_bridge
* Initial commit
* Contributors: Shashank Swaminathan


modelica_bridge_examples/src/modbridge_ex_teleop.cpp
/**
 * @brief Example teleop-based input for use with modbridge_node
 * @file modbridge_ex_teleop.cpp
 * 
 * Provides scaled joystick input to the modbridge_ex_controller node.
 * @author Shashank Swaminathan <sh.swami235@gmail.com>
 */
/*
 * Copyright 2018 Shashank Swaminathan.
 *
 * This file is part of the modelica_bridge_example package and subject to the license terms
 * in the top-level LICENSE file of the modelica_bridge repository.
 * https://github.com/ModROS/modelica_bridge/blob/master/LICENSE
 */

#include <ros/ros.h>
#include "modelica_bridge/ModComm.h"
#include <sensor_msgs/Joy.h>

ros::Publisher pub; ///< publisher for teleop input values

/**
 * Scales joystick input, and publishes to topic /modbridge_joy
 * Reads from the /joy topic, of the joy_node
 * @param [in] joy Message for joystick input
 * @see pub()
 * @return none
 */
void joyCallback(const sensor_msgs::Joy::ConstPtr& joy) {
  modelica_bridge::ModComm springVal;
  for(int i = 0; i < 8; i++) {
    springVal.data.push_back((i+1)*joy->axes[i]);
  }
  springVal.size = springVal.data.size();
  pub.publish(springVal);
}

/**
 * Initializes node, publisher, and subscriber. Spins to keep node running.
 * @see pub()
 * @see joyCallback()
 * @return 0
 */
int main(int argc, char **argv)
{
    ros::init(argc, argv, "modbridge_examples_teleop");
    ros::NodeHandle n;
    
    pub = n.advertise<modelica_bridge::ModComm>("modbridge_joy", 1);
    ros::Subscriber joy_sub = n.subscribe<sensor_msgs::Joy>("joy", 10, joyCallback);

    ros::spin();

    return 0;
}

modelica_bridge_examples/src/modbridge_ex_controller.cpp
/**
 * @brief Example controller for use with modbridge_node
 * @file modbridge_ex_controller.cpp
 * 
 * Receives teleop input from modbridge_ex_teleop, and feedback from model via modbridge_node.
 * Provides basic controller output to move feedback values to the input values.
 * @author Shashank Swaminathan <sh.swami235@gmail.com>
 */
/*
 * Copyright 2018 Shashank Swaminathan.
 *
 * This file is part of the modelica_bridge_example package and subject to the license terms
 * in the top-level LICENSE file of the modelica_bridge repository.
 * https://github.com/ModROS/modelica_bridge/blob/master/LICENSE
 */
#include "ros/ros.h"
#include "modelica_bridge/ModComm.h"

/** @def MAX_ARRAY
 *  @brief Defines the maximum size of internal data arrays
 */
#define MAX_ARRAY 100

ros::Publisher control_pub; ///< publisher for controller values

double springSetPoints[MAX_ARRAY] = {0.0}; ///< array to store control values

/**
 * Receives feedback from the model; calculates and publishes control values.
 * Reads from topic /model_values. Publishes to topic /control_values
 * Only reads MAX_ARRAY length of incoming data - rest is not stored.
 * @param [in] feedback_val ModComm message holding model feedback values
 * @see springSetPoints()
 * @see control_pub()
 * @see joyCallback()
 * @return none
 */
void controlCallback(const modelica_bridge::ModComm::ConstPtr& feedback_val) {

    modelica_bridge::ModComm control_val;
    for(int i = 0; i < feedback_val->size; i++) {
        control_val.data.push_back(5 * (springSetPoints[i] - feedback_val->data[i]));
    }
    control_val.size = control_val.data.size();

    control_pub.publish(control_val);
}

/**
 * Reads input from teleop node, stores as controller's set points
 * Reads on topic /modbridge_joy, of modelica_bridge_example's example teleop node
 * @param [in] inJoy
 * @see springSetPoints()
 * @see controlCallback()
 * @return none
 */
void joyCallback(const modelica_bridge::ModComm::ConstPtr& inJoy) {
    for(int i = 0; i<inJoy->size; i++) {
        springSetPoints[i] = inJoy->data[i];
    }
}

/**
 * Initializes node, pub, and subs. Spin to keep the node running
 * @see control_pub()
 * @see controlCallback()
 * @see joyCallback()
 * @return 0
 */
int main(int argc, char **argv)
{
    ros::init(argc, argv, "modbridge_examples_controller");
    ros::NodeHandle n;
    
    control_pub = n.advertise<modelica_bridge::ModComm>("control_values", 1);
    ros::Subscriber feedback_sub = n.subscribe("model_values", 1, controlCallback);
    ros::Subscriber joy_sub = n.subscribe("modbridge_joy", 1, joyCallback);

    ros::spin();

    return 0;
}

src/modbridge_node.cpp
/**
 * @brief ModBridge class and node
 * @file modbridge_node.cpp
 * 
 * Node for connecting Modelica to ROS via a tcp/ip interface.
 * 
 * This file is part of the modelica_bridge package and subject to the license terms
 * in the top-level LICENSE file of the modelica_bridge repository.
 * https://github.com/ModROS/modelica_bridge/blob/master/LICENSE
 * 
 * Copyright 2018
 * @author Shashank Swaminathan <sh.swami235@gmail.com>
 */

#include <ros/ros.h>
#include <modelica_bridge/ModComm.h>

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <stdio.h>
#include <unistd.h>
#include <netinet/in.h>
#include <signal.h>
#include <functional>

#include <stdlib.h>
#include <string.h>

/** @def MAX_BUF
    @brief Defines the maximum length of the character buffer
*/
/** @def MAX_ARRAY
    @brief Defines the maximum size of internal buffer arrays
*/
#define MAX_BUF 1024 // define maximum length of character buffer
#define MAX_ARRAY 256 // define maximum size of internal buffer arrays

/**
 * Class to act as relay between ROS and Modelica.
 * It initializes a tcp/ip socket to act as a bridge between ROS and Modelica. It 
 * can take character buffers up to 1024 bytes.
 */
class ModBridge
{
    public:
        /**
         * Defult constructor initializes receiving socket, subs and pubs.
         * No inputs.
         * @return none
         */
        ModBridge();
        /**
         * Keep program from closing.
         * Run callbacks for ROS nodes, and keep socket connection to Modelica alive.
         * No inputs.
         * @return none
         */
        void spin();

    private:
        /**
         * Callback for receiving controller values.
         * Reads from topic /control_values. 
         * Stores data into array of size MAX_ARRAY - only reads MAX_ARRAY amount of array slots.
         * @param[in] inVal ModComm message holding controller values.
         * @return none
         */
        void controllerCallback(const modelica_bridge::ModComm::ConstPtr& inVal);
        
        /**
         * Breaks up the incoming character buffer into data array.
         * Splits the character buffer into an array of tokens of data.
         * The first token is the size of the array.
         * @see buffer_()
         * @see ros_buffer_()
         * @see writeToBuffer()
         * @return none
         */
        void readFromBuffer();
        
        /**
         * Concatenates the control values into a character buffer.
         * @see readFromBuffer()
         * @see buffer_()
         * @see socket_buffer_()
         * @return none
         */
        void writeToBuffer();

        /**
         * Prints error message when interacting with socket.
         * @param [in] msg error message
         * @see error_check_()
         * @return none
         */
        void error(const char *msg);

        ros::NodeHandle nh_; ///< ROS node handle
        ros::Subscriber sub_; ///< subscriber to ROS controller values
        ros::Publisher pub_; ///< publisher for model feedback values

        int socket_fd_; ///< ROS socket binding
        int new_socket_fd_; ///< Modelica socket binding
        int port_num_; ///< Port number
        struct sockaddr_in serv_addr; ///< ROS socket address
        socklen_t client_len_; ///< Modelica socket size
        struct sockaddr_in client_address_; ///< Modelica socket address

        char buffer_[MAX_BUF]; ///< string buffer used to transmit information over the socket
        double socket_buffer_[MAX_ARRAY]; ///< buffer array for interaction with socket
        double ros_buffer_[MAX_ARRAY]; ///< buffer array for transmission via ROS
        int update_rate_; ///< ROS node maximum loop rate.

        int error_check_; ///< Parameter for checking errors on socket interaction.
};

ModBridge::ModBridge()
{
    sub_ = nh_.subscribe("control_values", 1, &ModBridge::controllerCallback, this);
    pub_ = nh_.advertise<modelica_bridge::ModComm>("model_values", 1);

    ros::NodeHandle nh_param("~");
    nh_param.param<int>("port_num", port_num_, 9091);
    nh_param.param<int>("update_rate", update_rate_, 20);

    socket_fd_ = socket(AF_INET, SOCK_STREAM, 0);
    if (socket_fd_ < 0) {
        error("ERROR opening socket");
    }

    bzero((char *) &serv_addr, sizeof(serv_addr));

    serv_addr.sin_family = AF_INET;
    serv_addr.sin_addr.s_addr = INADDR_ANY;
    serv_addr.sin_port = htons(port_num_);
    if (bind(socket_fd_, (struct sockaddr *) &serv_addr, sizeof(serv_addr)) < 0) {
        error("ERROR on binding");
    }
}

void ModBridge::spin() 
{
    ros::Rate loop(update_rate_);

    while(ros::ok())
    {
        modelica_bridge::ModComm outVal;

        listen(socket_fd_,5);
        client_len_ = sizeof(client_address_);

        new_socket_fd_ = accept(socket_fd_, (struct sockaddr *) &client_address_, &client_len_); 
        if (new_socket_fd_ < 0) {
            error("ERROR on accept");
        }

        error_check_ = read(new_socket_fd_, buffer_, MAX_BUF-1); 
        if (error_check_ < 0) {
            error("ERROR reading from socket");
        }

        readFromBuffer();
        writeToBuffer();

        error_check_ = write(new_socket_fd_, buffer_, strlen(buffer_));
        if (error_check_ < 0) {
            error("ERROR writing to socket");
        }

        close(new_socket_fd_);

        outVal.data.clear();
        outVal.size = ros_buffer_[0];
        for (int rosI = 1; rosI <= outVal.size; rosI ++) {
            outVal.data.push_back(ros_buffer_[rosI]);
        }
        pub_.publish(outVal);

        ros::spinOnce();
        loop.sleep();
    }
    close(socket_fd_);
}

void ModBridge::controllerCallback(const modelica_bridge::ModComm::ConstPtr& inVal)
{
    int i;
    for(i = 0; i < inVal->size; i++) {
        socket_buffer_[i] = inVal->data[i];
    }
}

void ModBridge::readFromBuffer()
{
    int j, i = 0;
    char *token[80];
    const char * delim = ",";

    token[0] = strtok(buffer_, delim);
    while (token[i] != NULL) {
        i++;
        token[i] = strtok(NULL, delim);
    }

    for (j=0; j<=i-1; j++) {
        ros_buffer_[j] = atof(token[j]);
    }
}

void ModBridge::writeToBuffer()
{
    buffer_[0] = '\0';
	char s1[100];
	for (int i = 0; i< sizeof(*socket_buffer_); i++)
	{
		sprintf(s1,"%f,",socket_buffer_[i]);
		strcat(buffer_,s1);
	}   
}

void ModBridge::error(const char *msg)
{
    perror(msg);
    exit(1);
}

int main(int argc, char** argv)
{
    ros::init(argc, argv, "modros_node");

    ModBridge modbridge;
    
    modbridge.spin();
    
    return 0;
}
